_transcripts/010-rust-1.40.0.md

```diff
@@ -0,0 +1,748 @@
+case at the very bottom, I think for structs, it's a less well known fact that
+you can match on structs, or that you can use a struct in a pattern context. So
+why don't you talk about, like, where structs are currently usable in pattern
+contexts and why you might want to do that.
+
+__Jon__: Sure. So the most common case for this is, you have a struct of some
+type, and you want to look at just one field of that struct. And one way that
+you can do that is, just— you have a type you have a variable `x` of type `foo`,
+and you do something like `x.field`, right, and that is one way to get at that
+that field of that struct. But there are a bunch of other times where you might
+want to— let's say, do a nested pattern match, right? So you want to look at a
+`foo` where it's `a` is a `bar` where the `bar`s `b` is something else, right?
+So you want to construct this nested pattern and you want to get, say, a
+reference to some field that's deeper down in that pattern. And maybe there's
+some enums or variants in between. And so it ends up being a pattern that's not
+just a direct field access.
+
+And in the past, as a library author, you sort of just have the choice of,
+either I make this field public, and then let people access it. But then if I
+add a field later, anyone who matches on it, that pattern is no longer going to
+compile, because it doesn't mention this new field that I've added. So adding a
+public field would be a non-backwards-compatible change. Right? If I had a type
+`foo` that had only one public field `a`, then previously— or in the current
+version, people could match on `foo` and then just write, sort of, `{ a }`, and
+then be able to access that field `a`. But if I add a public field `b` now, then
+that code that someone wrote with the current version of my library would no
+longer compile, because the compiler would say, this pattern doesn't match `foo`
+because `foo` has two fields, `a` and `b`. And so the way that you do pattern
+matching for these `non_exhaustive` types, or in general, when you do pattern
+matching the way that you should probably do it, is to include, for structs at
+least, the sort of `, ..` pattern. So this is saying "I'm going to match on
+these fields, and there are others that I want to ignore."
+
+__Ben__: Is that actually— does that work today? The `, ..` pattern?
+
+__Jon__: Yes, it does indeed. So this is a really common thing to use, even just
+internally in your code base, where— let's say you want references to two of the
+fields of `Foo` and you don't care about the others. You could write that as,
+like, `Foo { field1, field2, ` and then match on all the other fields as well,
+and then, just map them to underscore something. But that means that if you
+later changed `Foo` in your code base, you would have to update all of the
+pieces where `Foo` is matched on, and that's really annoying. And so instead, in
+the pattern you just write `Foo { field1, field2, .. }` and that means, there
```

Comment:
I'm not certain that `Foo { field1, field2, ` is the best way to write this part, because it's kind of a fragment, and the rest of the sentence explains what would come next.  Other ways this might be written:
- `Foo { field1, field2, ··· }` (unicode middle dot)
- `Foo { field1, field2, _, _ }`

The audio is at ~5:57.

Reply:
I think the current way it's set works pretty well. I'd keep it that way.